# NanoscopeAFM_ForceCurve_analysis


## _Very Brief Introduction_ 
These scripts are made with the intent of processing force curves obtained with AFM (https://en.wikipedia.org/wiki/Atomic_force_microscopy). A force curve describes the process of indentation of a biological sample (chromosomes in this particular case) by a sharp tip (https://www.nature.com/articles/s42254-018-0001-7). The curves record the force related to the indentation of the tip and its retraction from the sample, from them, multiple biophysical parameters related to the mechanical and structural properties of the sample can be retrieved.

### Workflow

Both _Nanoscope_converter.py_ , _nanoscope_CurvesContactPoint_determination.py_ and _nanoscope_CurveFit.py_ are used within the main analysis script, i.e., _**Chrm_IndentationAnalysis.py**_. In _**Chrm_IndentationAnalysis.py**_, a set of 1000+ curves (stored as raw binary files coming from the instrument) is parsed, converted into arrays which are then processed in order to find the "contact point" (i.e., where the first value of force > 0 is recorded); curves are also corrected for background noise, aligned in order to be compared etc.
After this, each curve is analysed, with the aim of extracting useful features from its shape (corresponding to biophysical parameters like the max pushing force, derivatives of force, energy of indentation etc...) that allow performing further clustering analyses. 


## _Contents_

Scripts for processing the force curves from Bruker AFMs. The folder "_Example_RawData_" contains examples of raw data used to test the scripts. The folder "_Example_figures_" contains the figures that are generated by the various scripts.

_**- Nanoscope_converter.py**:_ this script is a parser that reads the raw binary file containing the info on the force curve coming from the microscope and extract the information that allow plotting and processing the force curve (a representative figure of its output is present in the folder).

_**- nanoscope_CurvesContactPoint_determination.py**:_ this script contains two different functions that can be used to identify the contact point along the indenting force curve.

_The above-mentioned scripts are used within the other ones to process multiple curves and perform various analyses._

_**- nanoscope_CurveFit.py**:_ this script contains 2 different functions that allow processing and analysing the force curves. The first one is used to fit the indenting force curve with appropriate contact mechanics models; the second functions calculates the area under the indenting and retracting curve in order to extract information on the viscoelasticity of the sample (representative figures of its output is present in the folder).

_**- Chrm_IndentationAnalysis.py**:_ this script is used to analyse the force curves within the Example_RawData/Chromosomes folder, which refers to indentation experiments on multiple chromosomes. The script processes each curve (employing the previously mentioned scripts/functions) and extract useful parameters, like max force, derivatives, area under curve etc..., saving them into a .csv file for further clustering analyses.

_**- PCA_analysis.py**:_ reads the .csv file generated by Chrm_IndentationAnalysis.py and performs a PCA analysis in order to cluster curves having different shape and hence properties into distinct clusters. (a representative figure of its output is present in the folder)

_**- KMeans.py**:_ similar to PCA_analysis.py but performs KMean clustering on the data contained in the .csv file generated by Chrm_IndentationAnalysis.py (a representative figure of its output is present in the folder).